Add per-op MPS gradient tests and update skips

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84242

Follow up:
- ~Remove non-float dtypes from allow-list for gradients~
- ~Map dtypes to short-hand so there aren't so many lines, i.e. float16 should be f16.~
- ~There were a lot of linting issues that flake8 wouldn't format for me, so I reformatted with black. This makes the diff a little trickier to parse.~

Observations:
- there are entries in the allow-list that weren't there before 
- some forward that we previously passing now fail with requires_grad=True
- Because the allow list does not know about variants, a special skip was added for that in the block list

soulitzer committed:
Add per-op MPS gradient tests and update skips

[ghstack-poisoned]

albanD commented on test/test_mps.py:
```diff
@@ -6405,543 +6406,3244 @@ class TestConsistency(TestCase):
-                  'torch.int32',
-                  'torch.int64',
-                  'torch.uint8']}
+        "__getitem__": [
```
I feel like these either need to become blocklist (as there are more enabled than disabled) or we should split it in a different file to keep this one at a reasonable size.

soulitzer replied:
I've updated to use abbreviations for the dtypes (aka f16 instead of torch.float16) to reduce the size, perhaps it is more reasonable now?

albanD commented on test/test_mps.py:
```diff
@@ -6962,18 +9664,42 @@ def test_output_match(self, device, dtype, op):
+                device, dtype, requires_grad=(dtype.is_floating_point or dtype.is_complex)
+            )
+        cpu_samples = get_samples()
+        nb_samples = len(list(get_samples()))
```
Why?

soulitzer replied:
Whoops, forgot to clean this up

albanD commented on test/test_mps.py:
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
-        except Exception as e:
-            if not generate_new_truth:
-                raise e
-        else:
```
Note that this is not in the except, this is in the else that we append to the allow list.

soulitzer replied:
Yup, I later realized that too lol

albanD commented on test/test_mps.py:
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
+                all_forward_pass = False
+
+            # Skip the grad test if it is not part of the allow list
+            if not generate_new_truth and not run_grad_test:
```
This would better within the "Backward check" section I think.

albanD commented on test/test_mps.py:
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
+
+            # Skip the grad test if it is not part of the allow list
+            if not generate_new_truth and not run_grad_test:
+                if i == 0:
```
debug print

albanD commented on test/test_mps.py:
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
+                    for t in pytree.tree_flatten((mps_args, mps_kwargs))[0]
+                    if req_grad(t)
+                )
+                self.assertEqual(len(diff_cpu_out), len(diff_mps_out))
```
Also check input match as a sanity check?

albanD commented on test/test_mps.py:
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
+                continue
+
+            if not (dtype.is_floating_point or dtype.is_complex):
+                continue
```
This means that you never generate forward new truth for non-floating/complex dtypes?

albanD commented on test/test_mps.py:
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
+                self.NEW_ALLOW_LIST[op.name].append(str(dtype))
+            # We could write it only once. But I don't know how to detect that the current test is the last one
+            # So each test append to the dict and write it.
+            if i >= nb_samples - 1:
```
== ?

soulitzer replied:
Yep, but no longer needed

albanD reviewed (commented):
Sounds pretty good!

soulitzer committed:
Update on "Add per-op MPS gradient tests and update skips"


Follow up:
- Remove non-float dtypes from allow-list for gradients
- Map dtypes to short-hand so there aren't so many lines, i.e. float16 should be f16.
- There were a lot of linting issues that flake8 wouldn't format for me, so I reformatted with black. This makes the diff a little trickier to parse.

Observations:
- some forward that we previously passing now fail with requires_grad=True
- Because the allow list does not know about variants, a special skip was added for that in the block list


[ghstack-poisoned]

soulitzer committed:
Update on "Add per-op MPS gradient tests and update skips"


Follow up:
- ~Remove non-float dtypes from allow-list for gradients~
- ~Map dtypes to short-hand so there aren't so many lines, i.e. float16 should be f16.~
- ~There were a lot of linting issues that flake8 wouldn't format for me, so I reformatted with black. This makes the diff a little trickier to parse.~

Observations:
- there are entries in the allow-list that weren't there before 
- some forward that we previously passing now fail with requires_grad=True
- Because the allow list does not know about variants, a special skip was added for that in the block list


[ghstack-poisoned]

soulitzer committed:
Update on "Add per-op MPS gradient tests and update skips"


Follow up:
- ~Remove non-float dtypes from allow-list for gradients~
- ~Map dtypes to short-hand so there aren't so many lines, i.e. float16 should be f16.~
- ~There were a lot of linting issues that flake8 wouldn't format for me, so I reformatted with black. This makes the diff a little trickier to parse.~

Observations:
- there are entries in the allow-list that weren't there before 
- some forward that we previously passing now fail with requires_grad=True
- Because the allow list does not know about variants, a special skip was added for that in the block list


[ghstack-poisoned]

kulinseth commented:
Overall the commit looks good. Thanks for adding these new tests @soulitzer . I added the ciflow/mps labels to run the test_mps to make sure its green.

soulitzer committed:
Update on "Add per-op MPS gradient tests and update skips"


Follow up:
- ~Remove non-float dtypes from allow-list for gradients~
- ~Map dtypes to short-hand so there aren't so many lines, i.e. float16 should be f16.~
- ~There were a lot of linting issues that flake8 wouldn't format for me, so I reformatted with black. This makes the diff a little trickier to parse.~

Observations:
- there are entries in the allow-list that weren't there before 
- some forward that we previously passing now fail with requires_grad=True
- Because the allow list does not know about variants, a special skip was added for that in the block list


[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

malfet commented:
@pytorchbot merge -g
